qiskit/transpiler/passes/optimization/remove_final_reset.py

```diff
@@ -0,0 +1,37 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2023.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Remove reset when it is the final instruction on a qubit."""
+
+from qiskit.circuit import Reset
+from qiskit.dagcircuit import DAGOutNode
+from qiskit.transpiler.basepasses import TransformationPass
+
+
+class RemoveFinalReset(TransformationPass):
+    """Remove reset when it is the final instruction on a qubit wire."""
+
+    def run(self, dag):
+        """Run the RemoveFinalReset pass on `dag`.
+
+        Args:
+            dag (DAGCircuit): the DAG to be optimized.
+
+        Returns:
+            DAGCircuit: the optimized DAG.
+        """
+        resets = dag.op_nodes(Reset)
+        for reset in resets:
+            successor = next(dag.successors(reset))
+            if isinstance(successor, DAGOutNode):
+                dag.remove_op_node(reset)
```

Comment:
It'll probably be a bit more efficient to use the output_map in the dag and look at it's predecessors. This is because `dag.op_nodes()` iterates over all the instructions in the dag and in a typical circuit there will be far more instructions than qubits and clbits.

```suggestion
        for output_node in dag.output_map.values():
            if isinstance(output_node.wire, Qubit):
                pred = next(dag.predecessors(output_node))
                if isinstance(pred, DAGOpNode) and pred.op.name == "reset":
                    dag.remove_op_node(pred)
```

Reply:
Fixed in f358d46c33925f68598b67151caa1b8ba2de5631. I wonder if this could potentially be improved in the `RemoveResetInZeroState` pass too (I modeled my original code on this pass).

Reply:
Yeah, we can probably make a similar change to `RemoveResetInZeroState`. It looks like it hasn't had any meaningful updates since it was first introduced in #2139.